### DISCUSSION TITLE
BTCPayServer plugin's store

### DISCUSSION THREAD

**Discussion by NicolasDorier** [btcpayserver/btcpayserver#3961] [Ideas / Feature Requests]
@Kukks added plugin capabilities to BTCPay a while ago. While it has been experimental for a long time, it is now more exposed than before and actively used for many use case.
I think that the plugin success is incontested and that it is time that we invest more time into making plugin development and discovery easier, without breaking what has been done, while limiting maintenance impact.

## Current point I wish to solve:

First, most plugins link to `BTCPayServer.Abstractions`. 
This project is currently used for two reasons: 
* It provides an stable-ish interface that the plugins can use limiting the chances the plugin break during an update.
* While development of a plugin, it prevent circular dependency. (BTCPayServer project -> Plugin -> BTCPayServer.Abstractions)

I am not convinced this project is needed. I believe it is possible for plugins to be developped directly as part of BTCPayServer project. I will explain how next.

Second, the plugin page is doing a bunch of requests to github to display a sort of "Store" direclty in the instance. I think this doesn't scale, as plugins multiplies, we end up being rate limited by github, and the page is taking forever to open.

Third, the plugins are right now zipped binaries inside BTCPayServer.Plugins repository. This make it hard to know exactly who is publishing it and where the exact sources are.

## Solution proposed:

I propose the following direction:

Creating a website, `plugins.btcpayserver.org`, with a nice UI which will feature plugins, as well as instruction on how to install them on your server.
Ideally, we should be inspired by the UX of the wordpress store.
This website should be statically generated in a github page.

Changing the way our current BTCPayServer Plugins repository work: Rather than uploading the binaries directly into it, we would have a git module pointing to a specific fork of BTCPay Server.
BTCPayServer Plugins repository, would be responsible, via the CI, to pull the fork and extract the specific plugin folder from it.
Then BTCPay Server Plugins, as part of the CI, would insert the plugin in a specific version of BTCPayServer, then build then zip the binary by itself and upload it to some cloud storage. (or maybe just github)
Then, it would regenerate the `plugins.btcpayserver.org` to list this new plugin.

Finally, the current plugin page in BTCPay Server should instead link to the `plugins.btcpayserver.org` website rather than showing the available plugin directly.

In effect, developping a new plugin would be no different from developping on BTCPay Server as maintainers directly develop in the project itself and have access to all BTCPay's internal without the need of `BTCPayServer.Abstractions`.
Also, the BTCPayServer Plugins will be able able to check if plugins are compatible with new versions of BTCpay Server by just trying to compiled the extracted plugin source with the new BTCpay Server release branch. Removing one burden for the plugin developers.

**Comment by woutersamaey** [comment]
@NicolasDorier have you considered putting the plugins in separate repos? They can be added by the dev to the main project using a GIT submodule for easy development.
Because they are in their own repo, following changes, history, tags etc are easy.

Only thing needed is a GHA to "build" which would involve pulling the main project, copying the plugin to it, linking, building, etc. The binary can be extracted and committed to a `bin` or `dist` dir + added as a release.

I know @dennisreimann hates GIT submodules, but for me they have always worked well and there's no rebasing or merging with the main project development all the time. The GIT submodule doesn't need to be committed. You can commit it in your personal fork, or don't, which is also fine. The code is always in it's own repo, so no risky stahsing or whatever that may cause you to mess up and lose work.

If things break, the plugin can be fixed on a new branch for the BTCPay version involved, so that would also be a nice way of keeping older builds / versions there.

So, these are my 2 sats, but we can discuss on one of the calls if you like.

**Reply by woutersamaey** [reply]
Oh, and for the git submodule haters, there's this thing called git subtree too, which is similar but different. Which route you choose is a personal one as both are available in the context I described.

**Reply by dennisreimann** [reply]
I'm open for whatever improves the workflow and willing to revise that stance ;)

**Reply by woutersamaey** [reply]
I have prepared a build script that takes a plugin from it's own repo and builds it using the master project. Only thing I don't know yet is which files I need to extract at the end.... See https://github.com/storefront-bvba/btcpayserver-kraken-plugin/blob/master/build.sh

**Reply by NicolasDorier** [reply]
Having plugins in their own repo would be ideal, but we can't have that and at the same time give access to all classes in BTCPayServer.
Also, it makes them hard to develop, as how you can easily F5 debug it?

**Reply by woutersamaey** [reply]
@NicolasDorier actually this issue can be solved very easily and it is how we work at our company:
1. Clone the main btcpayserver project so you have everything
2. Add your plugin repo as a GIT submodule to the main project
3. Add your plugin as a reference, so it builds and runs, but don't commit those changes in the main project (you won't be committing there anyway)
4. Work on your plugin
5. Commit your work in the GIT submodule, changes are pushed to the separate repo
6. Set up some GHA to auto-build your plugin on push, along the lines of what I did in my build.sh script
7. Party

**Reply by woutersamaey** [reply]
If you haven't already, you may want to look into GIT submodules, what they are and how they work.

**Comment by mp21-org** [comment]
some points that where talked about on today's dev call, that we could discuss some more about:

1. why is installing a plugin such "an extremely dangerous operation"?
2. can we isolate plugin execution from the rest of BTCPay's Core so that Plugin developers are not able to DELETE FROM all our tables etc.? (SANDBOX Mode)
3. just a thought I had: Can we add additional execution environments for plugin developers? I.e. let's say I wanted to make a plugin in PHP - can I run that from inside btcpay's main container (just wanting to use the greenfield API and such)?
4. we should do basic linting and static code analysis of plugins prior to listing them on the plugin store
5. maybe introduce a kind of "badge" for approved versions of plugins that haven been manually reviewed (acknowleding that no one wants to take on the liability for any plugin they haven't created)
6. Plugin store should be integrated similiar to what Wordpress.com is doing: having a website and a showcase inside the App
7. Another approach to the central plugin store would be a system of linux distro like repositories that currate software for the user, than you can add through a URL (like EPEL or RPMFusion) via the UI

**Reply by AfolabiOlaoluwa** [reply]
I love these points. I was thinking these lines while reading through the discussion. It's valid! Confining Devs to a single language is not scalable. Plugins should have their store, submission, dynamism, and verifications. It should be removable if BTCPayServer user deems fit to uninstall a plugin at a click of a button.
